lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+
+    This information is orthogonal to that which is processed in `GitRef`. It describes
+    what operations are performed to *prepare* a git checkout for a spack `Stage`,
+    *after* fetching the configured ref.
+    """
+    submodules = None           # type: bool
+    submodules_delete = None    # type: Optional[List[str]]
+    get_full_repo = None        # type: bool
+
+    def __repr__(self):
+        return '{0}(submodules={1}, submodules_delete={2}, get_full_repo={3})'.format(
+            type(self).__name__,
+            self.submodules,
+            self.submodules_delete,
+            self.get_full_repo,
+        )
+
+    def __init__(self, submodules, submodules_delete, get_full_repo):
+        # type: (bool, Optional[List[str]], bool) -> None
+        self.submodules = submodules
+        self.submodules_delete = submodules_delete
+        self.get_full_repo = get_full_repo
+
+    @classmethod
+    def _extract_bool(cls, kwargs, name, default):
+        # type: (Any, str, bool) -> bool
+        value = kwargs.pop(name, default)
+        if isinstance(value, bool):
+            return value
+        raise InvalidGitFetchStageConfig(
+            'argument {0}={1!r} must be a bool'.format(name, value))
+
+    @classmethod
+    def from_version_directive(cls, kwargs):
+        # type: (Any) -> GitFetchStageConfiguration
+        try:
+            submodules = cls._extract_bool(kwargs, 'submodules', False)
+            submodules_delete = kwargs.pop('submodules_delete', None)
+            if not isinstance(submodules_delete, (list, type(None))):
+                raise InvalidGitFetchStageConfig(
+                    'argument submodules_delete={0!r} must be a list of str'
+                    .format(submodules_delete))
+            get_full_repo = cls._extract_bool(kwargs, 'get_full_repo', False)
+        except InvalidGitFetchStageConfig as e:
+            raise six.raise_from(  # type: ignore[attr-defined]
+                InvalidGitFetchStageConfig(
+                    'failed to parse {0} from kwargs {1}: {2}'
+                    .format(cls.__name__, kwargs, e)),
+                e,
+            )                   # type: ignore[func-returns-value]
+        return cls(submodules=submodules,
+                   submodules_delete=submodules_delete,
+                   get_full_repo=get_full_repo)
+
+
+class ConfiguredGit(object):
+    """Caches the version for the `git` executable, and adds compatibility args.
+
+    Use the `.from_executable()` method to perform this checking:
+
+        from spack.fetch_strategy import ConfiguredGit
+        from spack.util.Executable import which
+
+        git = ConfiguredGit.from_executable(which('git', required=True))
+        assert git.version == '2.31.1'
+
+    And note that the `__call__` method delegates to the inner executable from `which`:
+
+        assert git('--version', output=str) == 'git version 2.31.1'
+    """
+    git = None                 # type: Executable
```

Comment:
Does this even get instantiated multiple times? E.g., if it's a class variable changing for one class will change the others too.

Reply:
It begins as a class variable, but once `self.git` is assigned to within `__init__`, it becomes an instance variable! This `= None` method is the only way to annotate fields with mypy in python 2.

Reply:
It's also possible to avoid the field type declaration, and simply rely on assigning `self.git` within `__init__()` to make its type known to mypy, so it's slightly redundant to have this field type declaration at the top. I like using it because it shows me what fields the object has at a glance.